### PR TITLE
CFE-1047: Add the field "CapacityReservationGroupID" to "AzureMachineProviderSpec" in openshift/api 

### PR DESCRIPTION
**Pull request by anirudhAgniRedhat**
This PR adds the support of capacity reservation group in `AzureMachineProviderSpec`. A new field, `CapacityReservationGroupID`, is added. This ID is used while creating the corresponding Azure VM and the VM is added to the capacity reservation group.

The capacity reservation group should already exist. Once the VM is created and associated to the capacity reservation group it cannot be removed from the group. Similarly, if a VM is created without any association to a capacity reservation group then it cannot be associated to one later on. Thus, `CapacityReservationGroupID` field is made immutable.

### CODE REVIEW

File: machine/v1beta1/types_azureprovider.go [hunk 1]
```diff
@@ -145,6 +145,12 @@ type AzureMachineProviderSpec struct {
 	// This is useful for debugging software based launch issues.
 	// +optional
 	Diagnostics AzureDiagnostics `json:"diagnostics,omitempty"`
+	// CapacityReservationGroupID specifies the capacity reservation group resource id that should be
+	// used for allocating the virtual machine.
+	// It is optional but may not be changed once set.
```
Comment (arkadeepsen) [inline machine/v1beta1/types_azureprovider.go:150]: We don't need to add this line in the godoc.

Comment (anirudhAgniRedhat) [reply machine/v1beta1/types_azureprovider.go]: removed.

**Comment by arkadeepsen** [comment]
> This PR adds the support of capacity reservation group in `AzureMachineSpec`. 

`AzureMachineSpec` -> `AzureMachineProviderSpec`

File: machine/v1beta1/types_azureprovider.go [hunk 2]
```diff
@@ -145,6 +145,11 @@ type AzureMachineProviderSpec struct {
 	// This is useful for debugging software based launch issues.
 	// +optional
 	Diagnostics AzureDiagnostics `json:"diagnostics,omitempty"`
+	// CapacityReservationGroupID specifies the capacity reservation group resource id that should be
```
Comment (JoelSpeed) [inline machine/v1beta1/types_azureprovider.go:148]: Godoc should start with json tag not Go field name

File: machine/v1beta1/types_azureprovider.go [hunk 3]
```diff
@@ -145,6 +145,11 @@ type AzureMachineProviderSpec struct {
...
+	// used for allocating the virtual machine.
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="CapacityReservationGroupID is immutable once set"
```
Comment (JoelSpeed) [inline machine/v1beta1/types_azureprovider.go:150]: This validation must be spelt out in the godoc in prose

File: machine/v1beta1/types_azureprovider.go [hunk 4]
```diff
@@ -145,6 +145,11 @@ type AzureMachineProviderSpec struct {
...
+	// +optional
+	CapacityReservationGroupID string `json:"capacityReservationGroupID,omitempty"`
```
Comment (JoelSpeed) [inline machine/v1beta1/types_azureprovider.go:152]: What is the format of the reservation group ID? What characters are valid? What is the minimal and maximal length?

Comment (anirudhAgniRedhat) [reply machine/v1beta1/types_azureprovider.go:153]: @JoelSpeed I found this is the format for Capacity reservation Group ID 

I guess resourceGroupName and CapacityReservationGroupName may vary on differen Users.
`subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/capacityReservationGroups/<CapacityReservationGroupName>`

Also didn't found maximum and minimum length for ID in AzureDocs.

Ref: https://learn.microsoft.com/en-us/azure/virtual-machines/capacity-reservation-create?tabs=cli1%2Capi1%2Capi2#create-a-capacity-reservation-1

Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:153]: There's also a validation that it is a valid azure ID no?

Comment (anirudhAgniRedhat) [reply machine/v1beta1/types_azureprovider.go:153]: Yes, there  is a validation that the input must be string  having length more than 0,
Also the string should start with "/"

The keys which are used should be amoung "subscriptions", "providers" and "resourcegroups"
following by valid IDs respectively. 

the ideal format for ID is `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/capacityReservationGroups/<CapacityReservationGroupName>`

**Comment by anirudhAgniRedhat** [comment]
@JoelSpeed I will start working on its implementation from now onwards as we merge this PR.

File: machine/v1beta1/types_azureprovider.go [hunk 5]
```diff
@@ -145,6 +145,15 @@ type AzureMachineProviderSpec struct {
 	// This is useful for debugging software based launch issues.
 	// +optional
 	Diagnostics AzureDiagnostics `json:"diagnostics,omitempty"`
+	// capacityReservationGroupID specifies the capacity reservation group resource id that should be
+	// used for allocating the virtual machine.
+	// The field size should be greater than 0 and the field input must start with '/'.
+	// The input for capacityReservationGroupID must similar to '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/capacityReservationGroups/<CapacityReservationGroupName>'.
```
Comment (arkadeepsen) [inline machine/v1beta1/types_azureprovider.go:151]:
```suggestion
	// The input for capacityReservationGroupID must be similar to '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/capacityReservationGroups/{capacityReservationGroupName}'.
```

File: machine/v1beta1/types_azureprovider.go [hunk 6]
```diff
@@ -145,6 +145,15 @@ type AzureMachineProviderSpec struct {
...
+	// The keys which are used should be among 'subscriptions', 'providers' and 'resourcegroups' following by valid ID or names respectively.
```
Comment (arkadeepsen) [inline machine/v1beta1/types_azureprovider.go:152]:
```suggestion
	// The keys which are used should be among 'subscriptions', 'providers' and 'resourcegroups' followed by valid ID or names respectively.
```

**Review by arkadeepsen — COMMENTED** [review]
@JoelSpeed do these nits look good to be added?

**Comment by JoelSpeed** [comment]
Yeah lets get those added

Also, since we are close to the feature freeze date, I'd like to see the feature tested end to end by QE before we merge the API, else we risk shipping an API in one release with no implementation. Please link a WIP PR here once you have it

**Comment by arkadeepsen** [comment]
> Yeah lets get those added
> 
> Also, since we are close to the feature freeze date, I'd like to see the feature tested end to end by QE before we merge the API, else we risk shipping an API in one release with no implementation. Please link a WIP PR here once you have it

Sure @JoelSpeed

**Comment by anirudhAgniRedhat** [comment]
> Yeah lets get those added
> 
> Also, since we are close to the feature freeze date, I'd like to see the feature tested end to end by QE before we merge the API, else we risk shipping an API in one release with no implementation. Please link a WIP PR here once you have it

@JoelSpeed This is WIP PR for its implementation https://github.com/openshift/machine-api-provider-azure/pull/107

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

**Comment by anirudhAgniRedhat** [comment]
@JoelSpeed  The PR require`acknowledge-critical-fixes-only` label to merge it before 4.16 release If you feel we Should merge it Could you please add  the label?

**Comment by JoelSpeed** [comment]
Considering proximity to the release branch cut off, and the fact that TRT have had to introduce the critical fixes requirement. I'm inclined to say we should slip this feature to 4.17, do you see that causing an issue?

**Comment by anirudhAgniRedhat** [comment]
> Considering proximity to the release branch cut off, and the fact that TRT have had to introduce the critical fixes requirement. I'm inclined to say we should slip this feature to 4.17, do you see that causing an issue?

As this feature is anyway planed for 4.17, we should be good.

**Pull request merged**